### ISSUE TITLE
Union of type var and tuple of type var

### ISSUE THREAD

**Issue by gbelouze** [astral-sh/ty#2843]
### Summary

Not a typing guru, so not sure who is wrong between `ty`/`pyright`/`mypy` (or no-one and it is a choice of paradigm).
The following type checks for `pyright` and `mypy` but not `ty`

```python
from typing import TypeVar

T = TypeVar("T")


def as_resolution_tuple(resolution: T | tuple[T, ...]) -> tuple[T, ...]:
    if isinstance(resolution, tuple):
        return resolution
    return (resolution, resolution)
```

Is it expected ?

Note that the following type checks for `ty`

```python
from typing import TypeVar

T = TypeVar("T")


def as_resolution_list(resolution: T | list[T]) -> list[T]:
    if isinstance(resolution, list):
        return resolution
    return [resolution, resolution]
```

### Version

ty 0.0.17 (8cec85718 2026-02-13)

**Comment by carljm** [comment]
The lack of a diagnostic in the list case is a bug (at least relative to current intended/expected behavior) originating from this TODO comment: https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/src/types/generics.rs#L1028-L1037

Considering only the tuple case:

All of ty/mypy/pyright consider it possible that `T` itself is a tuple type which passes `isinstance(resolution, tuple)`, but they treat the effect of `isinstance(resolution, tuple)` differently. Mypy and pyright do not error because they consider the effect of `isinstance(resolution, tuple)` to be that `resolution` is `tuple[Unknown/Any, ...]`, which is assignable to `tuple[T, ...]` due to the forgiving nature of `Unknown/Any` -- they do not attempt to prove that `resolution` is _actually_ `tuple[T, ...]`.

ty is stricter, and considers the effect of the `isinstance` narrowing to be simply `tuple[object, ...]`, which is sound rather than forgiving, but is not assignable to `tuple[T, ...]`.

The interesting thing about this case is that if we can assume that the constraint solver will always prefer to unpack a tuple type from the caller when solving T (that is, a call to `as_resolution_tuple` with argument of type `tuple[int, ...]` will map to a return type of `tuple[int, ...]`, not `tuple[tuple[int, ...] | int, ...]`, then we could in principle record internally an invariant that `T` can never solve to a tuple, which would make this code safe.

Using ty-only features, this constraint can be expressed explicitly, which also makes this code pass without errors:

```py
from typing import TypeVar
from ty_extensions import Not

T = TypeVar("T", bound=Not[tuple[object, ...]])


def as_resolution_tuple(resolution: T | tuple[T, ...]) -> tuple[T, ...]:
    if isinstance(resolution, tuple):
        return resolution
    return (resolution, resolution)

def ft(x: tuple[int, ...]):
    reveal_type(as_resolution_tuple(x))  # revealed: tuple[int, ...]
```

**Comment by gbelouze** [comment]
Great! Thanks for the clear explanation and the solution for me :)
